relay-server/src/processing/transactions/mod.rs

```diff
@@ -0,0 +1,592 @@
+    RateLimited(RateLimits),
+}
+
+impl From<RateLimits> for Error {
+    fn from(value: RateLimits) -> Self {
+        Self::RateLimited(value)
+    }
+}
+
+impl OutcomeError for Error {
+    type Error = Self;
+
+    fn consume(self) -> (Option<Outcome>, Self::Error) {
+        let outcome = match &self {
+            Self::InvalidJson(_) => Outcome::Invalid(DiscardReason::InvalidJson),
+            Self::ProcessingFailed(e) => match e {
+                ProcessingError::InvalidTransaction => {
+                    Outcome::Invalid(DiscardReason::InvalidTransaction)
+                }
+                ProcessingError::EventFiltered(key) => Outcome::Filtered(key.clone()),
+                _other => {
+                    relay_log::error!(
+                        error = &self as &dyn std::error::Error,
+                        "internal error: trace metric processing failed"
+                    );
+                    Outcome::Invalid(DiscardReason::Internal)
+                }
+            },
+            Self::RateLimited(limits) => {
+                let reason_code = limits.longest().and_then(|limit| limit.reason_code.clone());
+                Outcome::RateLimited(reason_code)
+            }
+        };
+        (Some(outcome), self)
+    }
+}
+
+/// A processor for transactions.
+pub struct TransactionProcessor {
+    limiter: Arc<QuotaRateLimiter>,
+    geoip_lookup: GeoIpLookup,
+    #[cfg(feature = "processing")]
+    quotas_client: Option<AsyncRedisClient>,
+    reservoir_counters: ReservoirCounters,
+}
+
+impl Processor for TransactionProcessor {
+    type UnitOfWork = SerializedTransaction;
+
+    type Output = TransactionOutput;
+
+    type Error = Error;
+
+    fn prepare_envelope(
+        &self,
+        envelope: &mut ManagedEnvelope,
+    ) -> Option<Managed<Self::UnitOfWork>> {
+        let headers = envelope.envelope().headers().clone();
+
+        // The envelope might contain only a profile as a leftover after dynamic sampling.
+        // In this case, `transaction` is `None`.
+        let transaction = envelope
+            .envelope_mut()
+            .take_item_by(|item| matches!(*item.ty(), ItemType::Transaction));
+
+        // Attachments are only allowed if a transaction exists.
+        let attachments = match transaction {
+            Some(_) => envelope
+                .envelope_mut()
+                .take_items_by(|item| matches!(*item.ty(), ItemType::Attachment)),
+            None => smallvec::smallvec![], // no attachments allowed.
+        };
+
+        // A profile is only allowed if a transaction exists, or if it is marked as not sampled,
+        // in which case it is a leftover from a transaction that was dropped by dynamic sampling.
+        let profile = envelope.envelope_mut().take_item_by(|item| {
+            matches!(*item.ty(), ItemType::Profile) && (transaction.is_some() || !item.sampled())
+        });
+
+        if transaction.is_none() && profile.is_none() && attachments.is_empty() {
```

Comment:
If `transaction` is `None`, `attachments` is definitely empty.

```suggestion
        if transaction.is_none() && profile.is_none() {
```